moon/controller_excavator_round2.py

```diff
@@ -97,81 +129,215 @@ def vslam_reset_time(response):
-                        print (self.sim_time, self.robot_name, "%d: excavator: Distance: %f, index: %d in opposite direction, skipping" % (i, d, ind[i]))
-                        continue
-                    if -3*math.pi/4 < math.atan2(vol_list[ind[i]][1] - self.xyz[1], vol_list[ind[i]][0] - self.xyz[0]) < -math.pi/4:
-                        print (self.sim_time, self.robot_name, "%d: Too much away from the sun, shadow interference" % i)
-                        continue
+            # finish the turn no matter what
+            while True:
+                try:
+                    self.turn(normalizeAnglePIPI(dir_angle - self.yaw), timeout=timedelta(seconds=15))
+                    break
+                except:
+                    pass
 
-#                    if distance([0,0], vol_list[ind[i]]) > 35:
-#                        print (self.sim_time, self.robot_name, "%d: Too far from center" % i)
-#                        continue
+            # wait for hauler to start following, will receive osgar broadcast when done
+            while not self.hauler_ready:
+                try:
+                    self.wait(timedelta(seconds=1))
+                except:
+                    pass
 
+            self.send_request('external_command hauler_1 follow')
 
-                    try:
-                        print(self.sim_time, self.robot_name, "excavator: Pursuing volatile [%.1f,%.1f] at distance: %f" % (vol_list[ind[i]][0],vol_list[ind[i]][1], d))
-                        self.virtual_bumper = VirtualBumper(timedelta(seconds=3), 0.2) # radius of "stuck" area; a little more as the robot flexes
-                        with LidarCollisionMonitor(self, 1500): # some distance needed not to lose tracking when seeing only obstacle up front
-                            # turning in place probably not desirable because hauler behind may be in the way, may need to send it away first
-                            angle_diff = self.get_angle_diff(vol_list[ind[i]])
-                            self.turn(math.copysign(min(abs(angle_diff),math.pi/4),angle_diff), timeout=timedelta(seconds=15))
-                            self.go_to_location(vol_list[ind[i]], self.default_effort_level, offset=-0.5, timeout=timedelta(minutes=5), full_turn=True) # extra offset for sliding
-                            self.wait(timedelta(seconds=2)) # wait to come to a stop
-                            self.send_request('external_command hauler_1 approach %f' % self.yaw) # TODO: due to skidding, this yaw may still change after sending
-                            self.hauler_ready = False
-                            break
+            # reset again in case hauler was driving in front of excavator CANT RESET IF WE MOVED (TURNED)
+            # self.vslam_reset_at = None
+            # self.send_request('vslam_reset', vslam_reset_time)
+            #while self.vslam_reset_at is None or self.sim_time - self.vslam_reset_at < timedelta(seconds=3):
```

Comment:
again I am not sure if it is intentional to comment this code?

Reply:
yes, the way the 2 robots are started off, we cannot reset vslam again because the excavator turned meanwhile. If we started them differently, where the excavator stays put while hauler is going to its place, we can reset vslam which is helpful in case hauler drove in front of the camera and messed things up